cli/ops/fs.rs

```diff
@@ -146,6 +150,94 @@ fn op_open(
   }
 }
 
+fn open_helper(state: &State, args: Value) -> Result<(PathBuf, std::fs::OpenOptions), OpError> {
+  let args: OpenArgs = serde_json::from_value(args)?;
+  let path = Path::new(&args.path).to_path_buf();
+
+  let mut open_options = std::fs::OpenOptions::new();
+
+  if let Some(mode) = args.mode {
+    // mode only used if creating the file on Unix
+    // if not specified, defaults to 0o666
+    #[cfg(unix)]
+    {
+      use std::os::unix::fs::OpenOptionsExt;
+      open_options.mode(mode & 0o777);
+    }
+    #[cfg(not(unix))]
+    let _ = mode; // avoid unused warning
+  }
+
+  let options = args.options;
+  if options.read {
+    state.check_read(&path)?;
+  }
+
+  if options.write || options.append {
+    state.check_write(&path)?;
+  }
+
+  open_options
+    .read(options.read)
+    .create(options.create)
+    .write(options.write)
+    .truncate(options.truncate)
+    .append(options.append)
+    .create_new(options.create_new);
+
+  Ok((path, open_options))
+}
+
+fn op_open_sync(
+  isolate_state: &mut CoreIsolateState,
+  state: &State,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Result<Value, OpError> {
+  let resource_table = isolate_state.resource_table.clone();
+  let (path, open_options) = open_helper(state, args)?;
+  let std_file = open_options.open(path)?;
+  let tokio_file = tokio::fs::File::from_std(std_file);
+  let mut resource_table = resource_table.borrow_mut();
+  let rid = resource_table.add(
+    "fsFile",
+    Box::new(StreamResourceHolder::new(StreamResource::FsFile(Some((
+      tokio_file,
+      FileMetadata::default(),
+    ))))),
+  );
+  Ok(json!(rid))
+}
+
+use std::pin::Pin;
+use futures::Future;
+
+fn op_open_async(
+  isolate_state: &mut CoreIsolateState,
+  state: &State,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Pin<Box<dyn Future<Output = Result<Value, OpError>>>> {
```

Comment:
@piscisaureus could you take a look at `stateful_json_op_async` to turn this function into `async fn`?

Reply:
I've tried this:
```rust
async fn op_open_async(
  isolate_state: &mut CoreIsolateState,
  state: &State,
  args: Value,
  _zero_copy: &mut [ZeroCopyBuf],
) -> Result<Value, OpError> {
  let resource_table = isolate_state.resource_table.clone();
  let (path, open_options) = open_helper(&state, args)?;
  let tokio_file = tokio::fs::OpenOptions::from(open_options)
    .open(path)
    .await?;
  let mut resource_table = resource_table.borrow_mut();
  let rid = resource_table.add(
    "fsFile",
    Box::new(StreamResourceHolder::new(StreamResource::FsFile(Some((
      tokio_file,
      FileMetadata::default(),
    ))))),
  );
  Ok(json!(rid))
}
```
```rust
pub fn stateful_json_op_async<D, F>(&self, dispatcher: D) -> impl OpDispatcher
  where
    D: Fn(
      &mut deno_core::CoreIsolateState,
      &State,
      Value,
      &mut [ZeroCopyBuf],
    ) -> F,
    F: Future<Output = Result<Value, OpError>> + 'static
  {
  ... 
}
```
but I'm getting this error trying to compile:
```
▶ cargo build
   Compiling deno v1.2.3 (/Users/biwanczuk/dev/deno/cli)
error[E0271]: type mismatch resolving `for<'r, 's, 't0> <for<'_, '_, '_> fn(&mut deno_core::CoreIsolateState, &state::State, serde_json::Value, &mut [deno_core::ZeroCopyBuf]) -> impl futures::Future {ops::fs::op_open_async} as std::ops::FnOnce<(&'r mut deno_core::CoreIsolateState, &'s state::State, serde_json::Value, &'t0 mut [deno_core::ZeroCopyBuf])>>::Output == _`
  --> cli/ops/fs.rs:26:36
   |
26 |   i.register_op("op_open_async", s.stateful_json_op_async(op_open_async));
   |                                    ^^^^^^^^^^^^^^^^^^^^^^ expected bound lifetime parameter, found concrete lifetime

error: aborting due to previous error

For more information about this error, try `rustc --explain E0271`.
error: could not compile `deno`.
```